google/cloud/internal/oauth2_refreshing_credentials_wrapper.h

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_OAUTH2_REFRESHING_CREDENTIALS_WRAPPER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_OAUTH2_REFRESHING_CREDENTIALS_WRAPPER_H
+
+#include "google/cloud/status.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <string>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace oauth2_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Wrapper for refreshable parts of a Credentials object.
+ */
+class RefreshingCredentialsWrapper {
+ public:
+  struct TemporaryToken {
+    std::pair<std::string, std::string> token;
+    std::chrono::system_clock::time_point expiration_time;
+  };
+
+  template <typename RefreshFunctor>
+  StatusOr<std::pair<std::string, std::string>> AuthorizationHeader(
+      std::chrono::system_clock::time_point now,
+      RefreshFunctor refresh_fn) const {
+    if (IsValid(now)) {
```

Comment:
Maybe the code would be easier to understand if we renamed some of these helpers, consider:

```cc
  template <typename RefreshFunctor>
  StatusOr<std::pair<std::string, std::string>> AuthorizationHeader(
      std::chrono::system_clock::time_point now,
      RefreshFunctor refresh_fn) const {
  if (!NeedsRefresh(now)) return temporary_token_.token;

  auto new_token = refresh_fn();
  if (new_token) {
    temporary_token_ = *std::move(new_token);
    return temporary_token_.token;
  }
  // Beware ... "now" may no longer be "now" ...
  if (IsValid(now)) return temporary_token_.token;
  return std::move(new_token).status();
}
```

Naturally this requires changing the definition of `IsValid()`.